configure.ac

```diff
@@ -267,29 +254,104 @@ dnl --- enable_pkey
 enable_pkey=no
 
 case $target in
-     *s390x*)
-	  enable_pkey=yes
-	  ;;
-     *s390*)
-	  CFLAGS="$CFLAGS -m31 -DNO_PKEY"
-	  ;;
-     *ppc64* | *x86_64*)
-	  CFLAGS="$CFLAGS -DNO_PKEY"
-	  ;;
-     *ppc* | i*86*)
-	  CFLAGS="$CFLAGS -m32 -DNO_PKEY"
-	  ;;
-     *)
-	  CFLAGS="$CFLAGS -DNO_PKEY"
-	  ;;
+	*s390x*)
+		enable_pkey=yes
+	;;
+	*s390*)
+		CFLAGS="$CFLAGS -m31"
+		AC_DEFINE([NO_PKEY])
+	;;
+	*ppc64* | *x86_64*)
+		AC_DEFINE([NO_PKEY])
+	;;
+	*ppc* | i*86*)
+	  CFLAGS="$CFLAGS -m32"
+		AC_DEFINE([NO_PKEY])
+	;;
+	*)
+		AC_DEFINE([NO_PKEY])
+	;;
 esac
 AM_CONDITIONAL([NO_PKEY], [test "x$enable_pkey" != "xyes"])
 
+dnl Check and save whether the platform is AIX
+dnl Using the pthread library requires the _THREAD_SAFE macro to be set
+build_aix=no
+case "${host_os}" in
+	aix*)
+		build_aix=yes
+		AC_DEFINE([_ALL_SOURCE])
+		AC_DEFINE([_THREAD_SAFE])
+		with_systemd=no
+		with_libica=no
+		enable_ccatok=yes
+		enable_swtok=yes
+		enable_tpmtok=no
+		enable_icsftok=no
+		enable_ep11tok=no
+		echo "Disabling migration utilities on AIX: pkcscca pkcstok_migrate"
+		enable_pkcscca=no
+		enable_pkcstok_migrate=no
+		echo "Disabling the following unsupported features on AIX: systemd libica tpmtok icsftok ep11tok"
+		pkcsslotd_user=slotd
+		echo "Shortening username to $pkcsslotd_user due to platform restrictions"
+		function parse_map {
+			grep '^        ' $1 | tr -d ';' | tr -d ' ' > "$1".sym
+		}
+		parse_map opencryptoki.map
+		parse_map opencryptoki_tok.map
+		;;
+esac
+
+dnl Define custom variables
+lockdir=${localstatedir}/lock/opencryptoki
+logdir=${localstatedir}/log/opencryptoki
+
+dnl runstatedir is only defined starting from autoconf 2.70
+dnl this script has a prereq of 2.69, so instead build it from localstatedir
+if test "x$build_aix" = "xyes"; then
+	run_path=${localstatedir}/run
+else
+	run_path=/run
```

Comment:
How about using `${runstatedir}` instead of hard-coding it? Isn't this supposed to default to `/run`  ? Then distros could adjust it via `--runstatedir` during configure (like for other directories, too).

Thats available since autoconf 2.70

We might want to do something like this if we care about earlier autoconf  versions: 
https://gitlab.freedesktop.org/ljanda/NetworkManager/-/commit/921b7eb71699d39e0181d44a96500173397da957

Not sure if its worth the hassle, though....

Reply:
>How about using ${runstatedir} instead of hard-coding it? Isn't this supposed to default to /run ?

Nope, it still defaults to `/var/run`. The manual for the first autoconf [release](https://www.gnu.org/software/autoconf/manual/autoconf-2.70/html_node/Installation-Directory-Variables.html) with `runstatedir` confirms this as well. Quoting:-

>[..] By default, this is a subdirectory of localstatedir. 

..which is `${prefix}/var`. Since the sources have `/run` hard-coded regardless of the prefix specified during `configure`, I would prefer to keep that behaviour. Of course, that decision is up to you.

Reply:
OK, all right, then lets stay with hardcoded `/run` for now.

Is AIX also using systemd or init.d to start pkcsslotd? If so, then please note that these files also refer to `/run/opencryptoki/` for the PID file location:
https://github.com/opencryptoki/opencryptoki/blob/master/misc/pkcsslotd.in
https://github.com/opencryptoki/opencryptoki/blob/master/misc/pkcsslotd.service.in

Those files are processed during make by https://github.com/opencryptoki/opencryptoki/blob/master/misc/misk.mk so you might be able to incorporate the configured run dir into these files, too (if needed for AIX).

Reply:
No, it does not - it has its own equivalent of a stripped-down `systemd`. However, it is too trivial and does not handle modern daemon behaviours - such as `fork` followed by dropping privileges - well. Therefore, opencryptoki has not been integrated with this service - service starting must be done only by invoking the daemon through its absolute path.